platforms/android/library/src/main/java/io/element/android/wysiwyg/EditorEditText.kt

```diff
@@ -615,3 +637,61 @@ private fun KeyEvent.isMovementKey(): Boolean {
 private fun KeyEvent.isPrintableCharacter(): Boolean {
     return isPrintingKey || keyCode == KeyEvent.KEYCODE_SPACE
 }
+
+class EditorTextWatcher: TextWatcher {
+    private val nestedWatchers: MutableList<TextWatcher> = mutableListOf()
+
+    private val updateIsFromEditor: AtomicBoolean = AtomicBoolean(false)
+
+    var updateCallback: (CharSequence, Int, Int, CharSequence?) -> Unit = { _, _, _, _ -> }
+
+    fun addChild(watcher: TextWatcher) {
+        nestedWatchers.add(watcher)
+    }
+
+    fun removeChild(watcher: TextWatcher) {
+        nestedWatchers.remove(watcher)
+    }
+
+    fun inEditor(block: () -> Unit) {
+        updateIsFromEditor.set(true)
+        block()
+        updateIsFromEditor.set(false)
+    }
+
+    private var beforeText: CharSequence? = null
+
+    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
+        Timber.v("beforeTextChanged | text: \"$s\", start: $start, count: $count, after: $after")
+        if (!updateIsFromEditor.get()) {
+            beforeText = s?.subSequence(start, start + count)
+        }
+    }
+
+    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
+        Timber.v("onTextChanged | text: \"$s\", start: $start, before: $before, count: $count")
+        if (!updateIsFromEditor.get()) {
+            val newText = s?.subSequence(start, start + count) ?: ""
+            updateCallback(newText, start, start + before, beforeText)
+        }
+    }
+
+    override fun afterTextChanged(s: Editable?) {
+        Timber.v("afterTextChanged")
+        if (!updateIsFromEditor.get()) {
+            beforeText = null
+        }
+    }
+
+    fun notifyBeforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
+        nestedWatchers.forEach { it.beforeTextChanged(s, start, count, after) }
+    }
+
+    fun notifyOnTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
+        nestedWatchers.forEach { it.onTextChanged(s, start, before, count) }
+    }
+
+    fun notifyAfterTextChanged(s: Editable?) {
+        nestedWatchers.forEach { it.afterTextChanged(s) }
+    }
```

Comment:
I don't really like this solution, to be honest. I'd rather use the original overridden methods, but I couldn't find a good way to do it.